### PR TITLE
replace textstat_select with filter method

### PR DESCRIPTION
**Pull request by kbenoit**
Extends `dplyr::filter()` to data.frame-type returns from `textstat_*()` functions, with added arguments for `pattern`, `valuetype`, and `case_insensitive`.

Removes the `selection` argument, since it is inconsistent with `filter()` and since equivalent functionality can be returned using the negation operator or negative conditions on a regular expression.

### CODE REVIEW

**Comment by kbenoit** [comment]
See also #529.

**Comment by koheiw** [comment]
I think the best way is using `subset` as it was originally. It should have `select` and `subset`, but we just disallow using those and `patterns` at the same time.

File: R/textstat-methods.R [hunk 1]
```diff
@@ -1,71 +1,53 @@
...
-    id <- unlist(regex2id(pattern, x[[1]], valuetype, case_insensitive))
-    if (selection == 'keep') {
-        x <- x[id,]
-    } else {
-        x <- x[id * -1,]
+    # call dplyr filter, if ... arguments are supplied
+    if (length(list(substitute(...)))) {
+        if (!requireNamespace("dplyr", quietly = TRUE)) 
+            stop("You must install the dplyr package to use filter.textstat")
+        .data <- as.data.frame(dplyr::filter(dplyr::tbl_df(.data), ...))
```
Comment (hadley) [inline R/textstat-methods.R]: Why do you coerce back and forth to a tibble here?

File: R/textstat-methods.R [hunk 2]
```diff
@@ -1,71 +1,53 @@
...
-    attrs <- attributes(x)
+    attrs <- attributes(.data)
     
-    id <- unlist(regex2id(pattern, x[[1]], valuetype, case_insensitive))
-    if (selection == 'keep') {
-        x <- x[id,]
-    } else {
-        x <- x[id * -1,]
+    # call dplyr filter, if ... arguments are supplied
+    if (length(list(substitute(...)))) {
```
Comment (hadley) [inline R/textstat-methods.R]: I think this would be cleaner with a helper:

```R
ndots <- function(...) nargs()
```

File: R/textstat-methods.R [hunk 3]
```diff
@@ -1,71 +1,53 @@
...
+        if (!requireNamespace("dplyr", quietly = TRUE)) 
+            stop("You must install the dplyr package to use filter.textstat")
+        .data <- as.data.frame(dplyr::filter(dplyr::tbl_df(.data), ...))
     }
-    class(x) <- attrs$class
-    return(x)
+
+    # select on features if specified
+    if (!missing(pattern)) {
```
Comment (hadley) [inline R/textstat-methods.R]: Why are you not checking for `NULL` here?

File: R/textstat-methods.R (hunk 1 above)
Comment (kbenoit) [reply R/textstat-methods.R]: For some reason I had trouble accessing `dplyr::filter.data.frame`. But this solves it:
```r
dplyr:::filter.data.frame(.data, ...)
```

**Comment by kbenoit** [comment]
Thanks @hadley !

**Comment by koheiw** [comment]
Since `dplyr` is very popular it makes sense to add `filter()`, but conservatives like me prefer `subset()`. Why don't we have both?

**Comment by kbenoit** [comment]
Moved to [**quanteda.tidy**](https://github.com/quanteda/quanteda.tidy).

The `subset.data.frame()` function is supposed to do NSE on variable names, but here, `pattern`, `valuetype`, and `case_insensitive` are not variable names. What if one of them were? 

Although one could say that the same problem is true for `filter.textstat()`.

**Pull request closed without merging**